pkg/gofr/subscriber.go

```diff
@@ -24,39 +24,62 @@ func newSubscriptionManager(c *container.Container) SubscriptionManager {
 	}
 }
 
-func (s *SubscriptionManager) startSubscriber(topic string, handler SubscribeFunc) {
+func (s *SubscriptionManager) startSubscriber(ctx context.Context, topic string, handler SubscribeFunc) {
 	// continuously subscribe in an infinite loop
 	for {
-		msg, err := s.container.GetSubscriber().Subscribe(context.Background(), topic)
-		if msg == nil {
-			continue
-		}
-
-		if errors.Is(err, kafka.ErrConsumerGroupNotProvided) {
-			s.container.Logger.Errorf("cannot subscribe as consumer_id is not provided in configs")
+		select {
+		case <-ctx.Done():
+			s.container.Logger.Infof("shutting down subscriber for topic %s", topic)
 			return
-		} else if err != nil {
-			s.container.Logger.Errorf("error while reading from topic %v, err: %v", topic, err.Error())
-			continue
+		default:
+			err := s.handleSubscription(ctx, topic, handler)
+			if err != nil {
+				return
+			}
 		}
+	}
+}
 
-		ctx := newContext(nil, msg, s.container)
-		err = func(ctx *Context) error {
-			// TODO : Move panic recovery at central location which will manage for all the different cases.
-			defer func() {
-				panicRecovery(recover(), ctx.Logger)
-			}()
-
-			return handler(ctx)
-		}(ctx)
-
-		// commit the message if the subscription function does not return error
-		if err == nil {
-			msg.Commit()
-		} else {
-			s.container.Logger.Errorf("error in handler for topic %s: %v", topic, err)
-		}
+func (s *SubscriptionManager) handleSubscription(parentCtx context.Context, topic string, handler SubscribeFunc) error {
+	ctx := context.WithoutCancel(parentCtx)
```

Comment:
But then, you will never receive global context cancelation when app stops?

I'm unsure what you are trying to achieve with this change/PR

Reply:
global cancellation goes way up in call stack, this should be fine for handling single message. 
maybe i've been wrong thinking in terms of mqtt where context cancellation have no effect. but looking into kafka this may not be a good idea

Reply:
or maybe you can provide clearer concern about previous change
`	ctx, done := context.WithCancel(parentCtx)`
seems correct if you lookup kafka implementation

Reply:
Let me look again at the code

Reply:
The more I look, the more awkward it looks like.

Some subscribers are using context without cancel in Subscriber

I would have suggested to cancel the subscriber context. But it sounds a but strange.

Someone who have a clearer idea of the whole code base will have to look at your PR and tell what should be done.

The implementation depends on how long a subscribe is expect to last. Also, I don't get why the code keep subscribing and do jot consume a subscribed resource

Reply:
I don't think there's clear understanding among contributors given that this PR exists
Some code seems to be thrown in just to "have support of smth" (that I derived from some tickets here like "NATS support")
If you're interested we can do discord call to review some related code some time later

Reply:
We don't want to shut down the application when a subscriber cannot start. We kept in mind that a subscriber can be a part of a rest application where there are some endpoints registered, we did not want to make the server unreachable if there is a failure in case of a subscription. If only subscribers are registered, the server should not start and should shut down, but not otherwise. This was our design philosophy as we have seen such implementation frequently.